File: android/src/main/kotlin/com/zionmedianetwork/zmedia_player/BufferingHandler.kt
package com.zionmedianetwork.zmedia_player

import com.google.android.exoplayer2.ExoPlayer
import com.google.android.exoplayer2.DefaultLoadControl
import com.google.android.exoplayer2.LoadControl

/**
 * Handler for adaptive buffer management in ExoPlayer.
 *
 * Configures ExoPlayer's LoadControl to implement adaptive buffering
 * based on network conditions. Provides buffer health metrics for
 * predictive rebuffering.
 */
class BufferingHandler {

    companion object {
        // Default buffer configuration (milliseconds)
        private const val DEFAULT_MIN_BUFFER_MS = 2500
        private const val DEFAULT_MAX_BUFFER_MS = 50000
        private const val DEFAULT_TARGET_BUFFER_MS = 15000
        private const val DEFAULT_REBUFFER_MS = 5000

        // Fast startup configuration
        private const val FAST_MIN_BUFFER_MS = 1000
        private const val FAST_MAX_BUFFER_MS = 30000
        private const val FAST_TARGET_BUFFER_MS = 10000
        private const val FAST_REBUFFER_MS = 2500

        // Smooth playback configuration
        private const val SMOOTH_MIN_BUFFER_MS = 5000
        private const val SMOOTH_MAX_BUFFER_MS = 60000
        private const val SMOOTH_TARGET_BUFFER_MS = 20000
        private const val SMOOTH_REBUFFER_MS = 7500

        // Poor network configuration
        private const val POOR_MIN_BUFFER_MS = 10000
        private const val POOR_MAX_BUFFER_MS = 90000
        private const val POOR_TARGET_BUFFER_MS = 30000
        private const val POOR_REBUFFER_MS = 15000

        /**
         * Creates a LoadControl with default buffer configuration
         */
        fun createDefaultLoadControl(): LoadControl {
            return createLoadControl(
                minBufferMs = DEFAULT_MIN_BUFFER_MS,
                maxBufferMs = DEFAULT_MAX_BUFFER_MS,
                targetBufferMs = DEFAULT_TARGET_BUFFER_MS,
                rebufferMs = DEFAULT_REBUFFER_MS
            )
        }

        /**
         * Creates a LoadControl optimized for fast startup
         * Fast startup: Quick start (1s), maintain 10s buffer, max 30s
         */
        fun createFastStartupLoadControl(): LoadControl {
            return createLoadControl(
                minBufferMs = FAST_MIN_BUFFER_MS,        // 1000ms to start
                maxBufferMs = FAST_MAX_BUFFER_MS,        // 30000ms max
                targetBufferMs = FAST_TARGET_BUFFER_MS,  // 10000ms maintain
                rebufferMs = FAST_REBUFFER_MS            // 2500ms to resume
            )
        }

        /**
         * Creates a LoadControl optimized for smooth playback
         * Smooth: Slower start (5s), maintain 20s buffer, max 60s
         */
        fun createSmoothPlaybackLoadControl(): LoadControl {
            return createLoadControl(
                minBufferMs = SMOOTH_MIN_BUFFER_MS,        // 5000ms to start
                maxBufferMs = SMOOTH_MAX_BUFFER_MS,        // 60000ms max
                targetBufferMs = SMOOTH_TARGET_BUFFER_MS,  // 20000ms maintain
                rebufferMs = SMOOTH_REBUFFER_MS            // 7500ms to resume
            )
        }

        /**
         * Creates a LoadControl optimized for poor network conditions
         * Poor network: Large buffers (10s start, 30s maintain, 90s max)
         */
        fun createPoorNetworkLoadControl(): LoadControl {
            return createLoadControl(
                minBufferMs = POOR_MIN_BUFFER_MS,        // 10000ms to start
                maxBufferMs = POOR_MAX_BUFFER_MS,        // 90000ms max
                targetBufferMs = POOR_TARGET_BUFFER_MS,  // 30000ms maintain
                rebufferMs = POOR_REBUFFER_MS            // 15000ms to resume
            )
        }

        /**
         * Creates a LoadControl for live streaming with low latency
         * Live: Minimal buffers for low latency
         */
        fun createLiveStreamingLoadControl(targetLatencyMs: Int = 3000): LoadControl {
            return createLoadControl(
                minBufferMs = targetLatencyMs / 2,      // Half latency to start
                maxBufferMs = targetLatencyMs * 3,      // 3x latency max
                targetBufferMs = targetLatencyMs,       // Target latency maintain
                rebufferMs = targetLatencyMs            // Target latency to resume
            )
        }

        /**
         * Creates a LoadControl with custom buffer configuration
         *
         * Note: ExoPlayer parameter mapping:
         * - Dart targetBufferMs → ExoPlayer minBufferMs (maintain this buffer)
         * - Dart maxBufferMs → ExoPlayer maxBufferMs (don't exceed this)
         * - Dart minBufferMs → ExoPlayer bufferForPlaybackMs (start playback)
         * - Dart rebufferMs → ExoPlayer bufferForPlaybackAfterRebufferMs (resume after rebuffer)
         */
        fun createLoadControl(
            minBufferMs: Int,
            maxBufferMs: Int,
            targetBufferMs: Int,
            rebufferMs: Int
        ): LoadControl {
            return DefaultLoadControl.Builder()
                .setBufferDurationsMs(
                    targetBufferMs,   // ExoPlayer's minBufferMs - maintain this buffer during playback
                    maxBufferMs,      // ExoPlayer's maxBufferMs - maximum buffer allowed
                    minBufferMs,      // ExoPlayer's bufferForPlaybackMs - buffer needed to start
                    rebufferMs        // ExoPlayer's bufferForPlaybackAfterRebufferMs - buffer to resume
                )
                .build()
        }

        /**
         * Creates a LoadControl from Dart configuration map
         */
        fun createFromDartConfig(config: Map<String, Any>?): LoadControl {
            if (config == null) {
                return createDefaultLoadControl()
            }

            val minBufferMs = config["minBufferMs"] as? Int ?: DEFAULT_MIN_BUFFER_MS
            val maxBufferMs = config["maxBufferMs"] as? Int ?: DEFAULT_MAX_BUFFER_MS
            val targetBufferMs = config["targetBufferMs"] as? Int ?: DEFAULT_TARGET_BUFFER_MS
            val rebufferMs = config["rebufferMs"] as? Int ?: DEFAULT_REBUFFER_MS

            return createLoadControl(
                minBufferMs = minBufferMs,
                maxBufferMs = maxBufferMs,
                targetBufferMs = targetBufferMs,
                rebufferMs = rebufferMs
            )
        }
    }

    /**
     * Gets buffer health metrics from ExoPlayer
     */
    fun getBufferHealth(player: ExoPlayer?): Map<String, Any> {
        if (player == null) {
            return mapOf(
                "bufferedDurationMs" to 0,
                "currentPositionMs" to 0,
                "totalDurationMs" to 0,
                "downloadSpeed" to 0
            )
        }

        val bufferedPosition = player.bufferedPosition
        val currentPosition = player.currentPosition
        val duration = player.duration

        val bufferedDuration = (bufferedPosition - currentPosition).coerceAtLeast(0)

        // Get download speed estimate from bandwidth meter
        val downloadSpeed = try {
            // ExoPlayer's bandwidth estimate is in bits per second
            // Convert to bytes per second
            (player.currentTracks.groups.firstOrNull()?.mediaTrackGroup?.getFormat(0)?.bitrate ?: 0) / 8
        } catch (e: Exception) {
            0
        }

        return mapOf(
            "bufferedDurationMs" to bufferedDuration,
            "currentPositionMs" to currentPosition,
            "totalDurationMs" to if (duration >= 0) duration else 0,
            "downloadSpeed" to downloadSpeed
        )
    }

    /**
     * Determines if rebuffering is recommended based on current buffer status
     */
    fun shouldRebuffer(
        player: ExoPlayer?,
        minBufferMs: Int = DEFAULT_MIN_BUFFER_MS
    ): Boolean {
        if (player == null) return false

        val bufferedPosition = player.bufferedPosition
        val currentPosition = player.currentPosition
        val bufferedDuration = bufferedPosition - currentPosition

        return bufferedDuration < minBufferMs
    }
}

File: ios/Classes/BufferingHandler.swift
import Foundation
import AVFoundation

/**
 * Handler for adaptive buffer management in AVPlayer.
 *
 * Configures AVPlayer's preferredForwardBufferDuration to implement adaptive
 * buffering based on network conditions. Provides buffer health metrics for
 * predictive rebuffering.
 */
class BufferingHandler {

    // MARK: - Buffer Configuration Constants (in seconds)

    // Default buffer configuration
    private static let defaultMinBuffer: TimeInterval = 2.5
    private static let defaultMaxBuffer: TimeInterval = 50.0
    private static let defaultTargetBuffer: TimeInterval = 15.0
    private static let defaultRebuffer: TimeInterval = 5.0

    // Fast startup configuration
    private static let fastMinBuffer: TimeInterval = 1.0
    private static let fastMaxBuffer: TimeInterval = 30.0
    private static let fastTargetBuffer: TimeInterval = 10.0
    private static let fastRebuffer: TimeInterval = 2.5

    // Smooth playback configuration
    private static let smoothMinBuffer: TimeInterval = 5.0
    private static let smoothMaxBuffer: TimeInterval = 60.0
    private static let smoothTargetBuffer: TimeInterval = 20.0
    private static let smoothRebuffer: TimeInterval = 7.5

    // Poor network configuration
    private static let poorMinBuffer: TimeInterval = 10.0
    private static let poorMaxBuffer: TimeInterval = 90.0
    private static let poorTargetBuffer: TimeInterval = 30.0
    private static let poorRebuffer: TimeInterval = 15.0

    // MARK: - Buffer Configuration

    struct BufferConfig {
        let minBuffer: TimeInterval
        let maxBuffer: TimeInterval
        let targetBuffer: TimeInterval
        let rebuffer: TimeInterval

        static let `default` = BufferConfig(
            minBuffer: defaultMinBuffer,
            maxBuffer: defaultMaxBuffer,
            targetBuffer: defaultTargetBuffer,
            rebuffer: defaultRebuffer
        )

        static let fastStartup = BufferConfig(
            minBuffer: fastMinBuffer,
            maxBuffer: fastMaxBuffer,
            targetBuffer: fastTargetBuffer,
            rebuffer: fastRebuffer
        )

        static let smoothPlayback = BufferConfig(
            minBuffer: smoothMinBuffer,
            maxBuffer: smoothMaxBuffer,
            targetBuffer: smoothTargetBuffer,
            rebuffer: smoothRebuffer
        )

        static let poorNetwork = BufferConfig(
            minBuffer: poorMinBuffer,
            maxBuffer: poorMaxBuffer,
            targetBuffer: poorTargetBuffer,
            rebuffer: poorRebuffer
        )

        static func liveStreaming(targetLatency: TimeInterval = 3.0) -> BufferConfig {
            return BufferConfig(
                minBuffer: targetLatency / 2.0,
                maxBuffer: targetLatency * 3.0,
                targetBuffer: targetLatency,
                rebuffer: targetLatency
            )
        }

        static func fromDartConfig(_ config: [String: Any]?) -> BufferConfig {
            guard let config = config else {
                return .default
            }

            // Dart sends milliseconds, convert to seconds
            let minBufferMs = config["minBufferMs"] as? Int ?? Int(defaultMinBuffer * 1000)
            let maxBufferMs = config["maxBufferMs"] as? Int ?? Int(defaultMaxBuffer * 1000)
            let targetBufferMs = config["targetBufferMs"] as? Int ?? Int(defaultTargetBuffer * 1000)
            let rebufferMs = config["rebufferMs"] as? Int ?? Int(defaultRebuffer * 1000)

            return BufferConfig(
                minBuffer: TimeInterval(minBufferMs) / 1000.0,
                maxBuffer: TimeInterval(maxBufferMs) / 1000.0,
                targetBuffer: TimeInterval(targetBufferMs) / 1000.0,
                rebuffer: TimeInterval(rebufferMs) / 1000.0
            )
        }
    }

    // MARK: - Apply Buffer Configuration

    /**
     * Applies buffer configuration to AVPlayer.
     *
     * Note: AVPlayer uses preferredForwardBufferDuration for iOS 10+.
     * The value is a hint to the system; actual buffer size may vary.
     */
    static func applyBufferConfig(_ config: BufferConfig, to player: AVPlayer) {
        guard let playerItem = player.currentItem else {
            return
        }

        // Set preferred forward buffer duration
        // iOS will attempt to buffer this much content ahead of current time
        playerItem.preferredForwardBufferDuration = config.targetBuffer

        // Note: iOS doesn't expose min/max buffer configuration like ExoPlayer.
        // The system manages buffer sizes automatically based on network conditions.
        // We primarily control the target buffer duration.
    }

    /**
     * Applies buffer configuration from Dart config map.
     */
    static func applyDartConfig(_ dartConfig: [String: Any]?, to player: AVPlayer) {
        let config = BufferConfig.fromDartConfig(dartConfig)
        applyBufferConfig(config, to: player)
    }

    // MARK: - Buffer Health Monitoring

    /**
     * Gets buffer health metrics from AVPlayer.
     */
    static func getBufferHealth(from player: AVPlayer?) -> [String: Any] {
        guard let player = player,
              let playerItem = player.currentItem else {
            return [
                "bufferedDurationMs": 0,
                "currentPositionMs": 0,
                "totalDurationMs": 0,
                "downloadSpeed": 0
            ]
        }

        // Get current time
        let currentTime = CMTimeGetSeconds(playerItem.currentTime())
        let currentPositionMs = Int(currentTime * 1000)

        // Get total duration
        let duration = CMTimeGetSeconds(playerItem.duration)
        let totalDurationMs = duration.isFinite ? Int(duration * 1000) : 0

        // Calculate buffered duration
        let bufferedDuration = getBufferedDuration(from: playerItem, currentTime: currentTime)
        let bufferedDurationMs = Int(bufferedDuration * 1000)

        // Estimate download speed (simplified)
        let downloadSpeed = estimateDownloadSpeed(from: playerItem)

        return [
            "bufferedDurationMs": bufferedDurationMs,
            "currentPositionMs": currentPositionMs,
            "totalDurationMs": totalDurationMs,
            "downloadSpeed": downloadSpeed
        ]
    }

    /**
     * Calculates buffered duration from AVPlayerItem's loaded time ranges.
     */
    private static func getBufferedDuration(from playerItem: AVPlayerItem, currentTime: TimeInterval) -> TimeInterval {
        let loadedTimeRanges = playerItem.loadedTimeRanges

        guard !loadedTimeRanges.isEmpty else {
            return 0
        }

        // Find the time range that contains or follows the current time
        var bufferedDuration: TimeInterval = 0

        for value in loadedTimeRanges {
            let timeRange = value.timeRangeValue
            let start = CMTimeGetSeconds(timeRange.start)
            let duration = CMTimeGetSeconds(timeRange.duration)
            let end = start + duration

            // If current time is within this range
            if currentTime >= start && currentTime <= end {
                bufferedDuration = end - currentTime
                break
            }
            // If this range is ahead of current time
            else if start > currentTime {
                bufferedDuration = end - currentTime
                break
            }
        }

        return max(0, bufferedDuration)
    }

    /**
     * Estimates download speed from AVPlayerItem's access log.
     * Returns bytes per second.
     */
    private static func estimateDownloadSpeed(from playerItem: AVPlayerItem) -> Int {
        guard let accessLog = playerItem.accessLog(),
              let lastEvent = accessLog.events.last else {
            return 0
        }

        // observedBitrate is in bits per second
        // Convert to bytes per second
        let bitsPerSecond = lastEvent.observedBitrate
        if bitsPerSecond > 0 {
            return Int(bitsPerSecond / 8)
        }

        // Fallback: use indicated bitrate
        let indicatedBitrate = lastEvent.indicatedBitrate
        if indicatedBitrate > 0 {
            return Int(indicatedBitrate / 8)
        }

        return 0
    }

    /**
     * Determines if rebuffering is recommended based on current buffer status.
     */
    static func shouldRebuffer(player: AVPlayer?, minBuffer: TimeInterval = defaultMinBuffer) -> Bool {
        guard let player = player,
              let playerItem = player.currentItem else {
            return false
        }

        let currentTime = CMTimeGetSeconds(playerItem.currentTime())
        let bufferedDuration = getBufferedDuration(from: playerItem, currentTime: currentTime)

        return bufferedDuration < minBuffer
    }
}
